src/pid-file/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+    ///
+    /// If the file is already locked by another process, it returns
+    /// `Error::AlreadyRunning`.
+    pub fn open<P>(path: P) -> Result<PidFile, Error>
+    where
+        P: AsRef<Path>,
+    {
+        PidFile::open_with(path, Permissions::from_mode(0o600))
+    }
+
+    /// Like [`open`](PidFile::open), but opens the file with the specified
+    /// permissions rather than 0600.
+    pub fn open_with<P>(path: P, permissions: Permissions) -> Result<PidFile, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let path_cstring = CString::new(path.as_ref().as_os_str().as_bytes())?;
+        let mut old_pid: libc::pid_t = -1;
+        let mode: libc::mode_t = permissions
+            .mode()
+            .try_into()
+            .expect("file permissions not valid libc::mode_t");
+        let f = unsafe { pidfile_open(path_cstring.as_ptr(), mode, &mut old_pid) };
+        if !f.is_null() {
+            let r = unsafe { pidfile_write(f) };
+            if r == 0 {
+                Ok(PidFile(f))
+            } else {
+                Err(Error::Io(io::Error::last_os_error()))
+            }
+        } else {
+            let err = io::Error::last_os_error();
+            if err.kind() == io::ErrorKind::AlreadyExists {
+                Err(Error::AlreadyRunning {
+                    pid: (old_pid != -1).then(|| old_pid),
+                })
+            } else {
+                Err(Error::Io(err))
+            }
+        }
+    }
+
+    /// Closes the PID file and removes it from the filesystem.
+    pub fn remove(mut self) -> Result<(), Error> {
+        let r = unsafe { pidfile_remove(self.0) };
+        self.0 = ptr::null_mut();
+        if r == 0 {
+            Ok(())
+        } else {
+            Err(Error::Io(io::Error::last_os_error()))
+        }
+    }
+}
+
+impl Drop for PidFile {
+    fn drop(&mut self) {
+        if self.0 != ptr::null_mut() {
+            unsafe { pidfile_remove(self.0) };
+        }
+    }
+}
+
+/// A PID file-related error.
+#[derive(Debug)]
+pub enum Error {
+    /// An I/O error occurred.
+    Io(io::Error),
+    /// The provided path had embedded null bytes.
+    Nul(NulError),
+    /// Another process already has the lock on the requested PID file.
+    AlreadyRunning {
+        /// The PID of the existing process, if it is known.
+        pid: Option<i32>,
+    },
+}
+
+impl From<NulError> for Error {
+    fn from(e: NulError) -> Error {
+        Error::Nul(e)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Io(e) => write!(f, "unable to open PID file: {}", e),
+            Error::Nul(e) => write!(f, "PID file path contained null bytes: {}", e),
+            Error::AlreadyRunning { pid } => write!(
+                f,
+                "process already running (PID: {})",
+                pid.display_or("<unknown>")
+            ),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
```

Comment:
Maybe a simple test (like you have for Service)?

Reply:
Done.